src/D2L.CodeStyle.Analyzers/ApiUsage/Events/EventHandlerLoaderTypesAnalyzer.cs

```diff
@@ -0,0 +1,181 @@
+						eventHandlerAttributeType,
+						genericRegisterMethods
+					),
+					SyntaxKind.InvocationExpression
+				);
+		}
+
+		private void AnalyzeMethodInvocation(
+				SyntaxNodeAnalysisContext context,
+				InvocationExpressionSyntax invocation,
+				INamedTypeSymbol eventAttributeType,
+				INamedTypeSymbol eventHandlerAttributeType,
+				IImmutableSet<ISymbol> genericRegisterMethods
+			) {
+
+			ISymbol expessionSymbol = context.SemanticModel
+				.GetSymbolInfo( invocation.Expression )
+				.Symbol;
+
+			if( expessionSymbol.IsNullOrErrorType() ) {
+				return;
+			}
+
+			if( !( expessionSymbol is IMethodSymbol methodSymbol ) ) {
+				return;
+			}
+
+			if( !methodSymbol.IsGenericMethod ) {
+				return;
+			}
+
+			if( !genericRegisterMethods.Contains( methodSymbol.OriginalDefinition ) ) {
+				return;
+			}
+
+			ITypeSymbol eventTypeSymbol = methodSymbol.TypeArguments[ 0 ];
+			if( eventTypeSymbol.IsNullOrErrorType() ) {
+				return;
+			}
+
+			ITypeSymbol eventHandlerSymbol = methodSymbol.TypeArguments[ 1 ];
+			if( eventHandlerSymbol.IsNullOrErrorType() ) {
+				return;
+			}
+
+			InspectEventType( context, invocation, eventAttributeType, eventTypeSymbol );
+			InspectEventHandlerType( context, invocation, eventHandlerAttributeType, eventHandlerSymbol );
+		}
+
+		private static void InspectEventType(
+				SyntaxNodeAnalysisContext context,
+				InvocationExpressionSyntax invocation,
+				INamedTypeSymbol eventAttributeType,
+				ITypeSymbol eventTypeSymbol
+			) {
+
+			bool hasAttr = eventTypeSymbol
+				.GetAttributes()
+				.Any( attr => attr.AttributeClass.Equals( eventAttributeType ) );
+
+			if( hasAttr ) {
+				return;
+			}
+
+			if( eventTypeSymbol.TypeKind == TypeKind.TypeParameter ) {
+				return;
+			}
+
+			Diagnostic diagnostic = Diagnostic.Create(
+					Diagnostics.EventTypeMissingEventAttribute,
+					invocation.GetLocation(),
+					eventTypeSymbol.ToDisplayString()
+				);
+
+			context.ReportDiagnostic( diagnostic );
+		}
+
+		private static void InspectEventHandlerType(
+				SyntaxNodeAnalysisContext context,
+				InvocationExpressionSyntax invocation,
+				INamedTypeSymbol eventHandlerAttributeType,
+				ITypeSymbol eventHandlerSymbol
+			) {
+
+			bool hasAttr = eventHandlerSymbol
+				.GetAttributes()
+				.Any( attr => attr.AttributeClass.Equals( eventHandlerAttributeType ) );
+
+			if( hasAttr ) {
+				return;
+			}
+
+			if( eventHandlerSymbol.TypeKind == TypeKind.TypeParameter ) {
+				return;
+			}
+
+			Diagnostic diagnostic = Diagnostic.Create(
+					Diagnostics.EventHandlerTypeMissingEventAttribute,
+					invocation.GetLocation(),
+					eventHandlerSymbol.ToDisplayString()
+				);
+
+			context.ReportDiagnostic( diagnostic );
+		}
+
+		private static IEnumerable<ISymbol> GetGenericRegisterMethods( Compilation compilation ) {
+
+			INamedTypeSymbol registryType = compilation.GetTypeByMetadataName( IEventHandlerRegistryFullName );
+			if( registryType.IsNullOrErrorType() ) {

```

Comment:
Doesn't make sense for `GetTypeByMetadataName` (here and elsewhere) to return an error type and it definitely won't because of this line: http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/AssemblySymbol.cs,584 Not a big deal I just like to thing through these more precisely.

Reply:
I'll do a pass and replace this type of thing broader with an TryGet extension method